Metacluster gui

y2kbugger opened the pull request:
## Slack
Full development notes in slack: https://toptal.slack.com/archives/C02762FAZB6/p1626918772008500

## Launching
I have added jupyerlab to the requirements file because it gives more horizontal space to use the GUI, to try this for yourself, after starting jupyter you can change the url to `lab` instead of `tree`.

e.g. `http://127.0.0.1:8888/tree/` -> `http://127.0.0.1:8888/lab/`

## Usage
- To select one or more clusters, click anywhere on either heatmap. Clicking a selected cluster will deselect it.
- To remap the selected clusters, right click anywhere on either heatmap.
- The bars at the top show the relative size of each cluster.
- If you remap all cluster in a metacluster, the metacluster is now *gone*
- You can click new metacluster to add current selection to a new metacluster
- You final mapping will be in an output file in the same directory as the input files

![image](https://user-images.githubusercontent.com/6101677/126728358-788b1128-6ecf-45c2-ad39-348962a772b6.png)
![image](https://user-images.githubusercontent.com/6101677/126728449-9121e246-3ba0-4393-bef8-449ef47ca8e9.png)

## Backlog

- Todo Next
  - barchart for aggregate pixelsize of metaclusters
  - order rows by hcluster ordering | 5
  - metacluster name labels | 2
  - change names | 5

- bug
  - ensure stable sorts
- feature
  - hide/show some markers
  - undo | 8
  - alternatively add a prompt for (are you sure? before remapping) | 2
  - add statistics to highlight potential remaps | 5 -

ngreenwald reviewed (changes requested):
Looks great so far! Couple suggestions for functionality
1. Add a y axis label for barchart counts on top of each column
2. Add a rowname for the metacluster row and selection row

Once you're happy with the functionality, some code/documentation suggestions:
1. Please pull out all of the functions into a separate .py file that gets imported at the top
2. Add a markdown cell with a brief description of how to work with the GUI
3. Change the column name for clustering into variable (instead of hardcoding `hCluster_cap`)
4. Input validation for supplied data to make sure it's the appropriate type, has the required column names, etc

y2kbugger commented:
Thanks for the feedback I've added the functionality requests to the todo, they should be straightforward.

@ngreenwald I've added my "maintainability" backlog above since it seems to be appreciated. I laughed when I compared our lists, our first three were exactly the same:

## design/maintainability
-   **move object code to separate  file**
-   **remap hCluster_cap upon loading, allow to pass alternate name as to not hardcode**
-   **Add usage docs**
-   Update/finalize maintenance docs/comments
-   could reduce recalculation with caching of properties
    -   e.g. metaclusters, expire on mapping changes
-   put all references that peer into mcd behind methods, e.g. columns, etc
-   plot only dummy data in the create gui, move as much as possible to "update_gui" (DRY)

cliu72 reviewed (commented):
Looks awesome! Agree with everything that's been said, only other suggestion is to add the dendrogram/hierarchical clustering of the markers.

y2kbugger commented:
Are you all fine with leaving jupyterlab as a required dependency? If so, I can think of two ways forward:

1. Change default launcher to `jupyter lab` (instead of `jupyter notebook`)
    - I'd need some help testing that all other notebooks continue to work as expected.
2. I'll just leave my instructions in a markdown cell about how to open Jupyter Lab using the URL address bar.

Again the reason for this is because it gives more horizontal room for the Notebook based GUI.

y2kbugger commented:
Regarding the ability to give metaclusters human readable names:

Are the names needed in an output, or should they only live as long as you are running the interactive session?

y2kbugger commented:
## 2021-07-25
- pixel count data labels | 3
- move code to imported python package
- move usage docs to example notebook
- row labels for metacluster and selection | 1
- Ensure color stability of metaclusters
- Make metacluster plot via cluster number rather than index

![image](https://user-images.githubusercontent.com/6101677/126914647-078c14f3-a6f9-4fbe-9a09-b3d006692fd4.png)

ngreenwald commented:
I tried jupyter lab for like 2 minutes and couldn't get the notebook to work, so went back to regular. I don't think having the scroll bar was a big deal, so my preference would be to keep things simple. Up to you Candace. 

Re: cluster names, it would definitely be useful if the user-generated names replaced the current default names in the output.

cliu72 commented:
I was able to get jupyter lab working after rebuilding the docker without issue. I do think being able to view both heatmaps together without scrolling is a big advantage, since we want to see how the metacluster heatmap changes after re-assigning something, and it may get annoying to scroll back and forth. If it's too much trouble to switch the default to jupyter labs (Noah, not sure what the error is for you), we can just leave the instructions at the top of this notebook. If a user can't get jupyter labs to work, they can always just deal with the scrolling in jupyter notebook.

For the cluster names, it would be great if you could output the user-generated names, so the final output is a file with 100 rows and 3 columns: cluster number (1-100), metacluster number, and metacluster name.

Also not sure if I missed this somewhere in your notes, but it would be great if the default colors were blue-white-red (easier to visualize z-scores like that).

y2kbugger commented:
## Jupyter lab 
either way is fine with me. If we're trying to keep it simple for the user,  I think chances are good it would be safe to launch it as Jupyter Lab from the start docker script. I think in general Jupyter Lab is the preferred way of using notebooks when working with multiple open notebooks.

Official docs:

> ![image](https://user-images.githubusercontent.com/6101677/127030082-d3ea6561-86ce-413a-9bd2-dfb6125ca2af.png)

## Saving names
Got it, I'll design the naming so with saving in mind.

## Color scheme
Blue and white is already the default, I'll make sure to document the parameters better in the usage cell. I noticed that I get a better sense of magnitude with a non-divergent color scheme (especially while tweaking zscore in realtime with slider), so I overrode the default (with "viridis") for myself when developing and forgot to remove. Let me know what you think. I can make any Matplotlib scheme the default.

> ![image](https://user-images.githubusercontent.com/6101677/127030700-bf44ff38-aadf-4e22-85c5-db804a57f892.png)

PS. Thinking about it, it could be that the divergent heatmap is not mapping blue-white-red to min-zero-max, but instead giving a linear mapping of blue-white-red across min-max. Let me know if you want me to spend some time looking into this.

## Marker Dendrogram
PPS. Still cooking about ideas for how to best show dendrogram for markers. I think i could draw the tree. I could also just order the rows and give indication of hierarchy in the label name (uglier, might be less effort if drawing tree to the independent mpl ax  turns out to be hard). Main decision to be made is: should this part of pipeline rerun clusterering or take the heirarchy as an input. I think you are using some built in clustering from seaborn, correct me if wrong. I can look into how to save out the dendrogram structure if you want.

ngreenwald commented:
Sure thing, sounds good. Can you update the docker script and I'll check out the other notebooks to make sure nothing weird happened?

y2kbugger commented:
> Sure thing, sounds good. Can you update the docker script and I'll check out the other notebooks to make sure nothing weird happened?

Done, should be working by default after pull and rebuild.

ngreenwald commented:
This is the [seaborn clustermap](https://seaborn.pydata.org/generated/seaborn.clustermap.html) function candace was referring to. No issues with our other notebooks, I think switching to Jupyter Lab is totally fine.

y2kbugger commented:
Great I'll leave that in.

On Wed, Jul 28, 2021 at 5:41 PM Noah F. Greenwald ***@***.***>
wrote:

> This is the seaborn clustermap
> <https://seaborn.pydata.org/generated/seaborn.clustermap.html> function
> candace was referring to. No issues with our other notebooks, I think
> switching to Jupyter Lab is totally fine.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/angelolab/ark-analysis/pull/432#issuecomment-888640378>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABORVLO5ILZPXJLDAZVWHSLT2B2RDANCNFSM5A3DZN4Q>
> .
>

y2kbugger commented:
## 2021-07-28
- Setup local envs to enable linting/refactoring in vscode
- Clean up accidentally committed ipynb in wrong folder
- Working through lint

## 2021-07-29
- Fixup and put in pull request for enhanced Jupyter launching
- fix linting errors | 1
- Started adding unit and functional pytests
  - Add pure constructors for testing with simple dummy data

y2kbugger commented:
## 2021-07-31

- add test cases | 2
- remap hCluster_cap upon loading, allow to pass alternate name as to not hardcode
- Add backend for metacluster displaynames

y2kbugger commented:
## 2021-08-05
- Add unit tests for GUI w/ Virtual clicks

y2kbugger commented:
## 2021-08-09

- Finish unit tests for GUI
- add tests for distinct colormap
- add tests for throttle
- Clean up asyncio throttle implementation
- Achieve acceptable test coverage
- Start bar charts for metacluster totals

y2kbugger commented:
**2021-08-12**
- Test for weighted average/metaclustering
- Implement colormap normalizer to keep the dynamic range good, but keep the diverging heatmap centered about 0.
- Add mc names to output file
- Optimize frequency out outputfile saves
- Update usage documentation in Notebook

y2kbugger commented:
## 2021-08-17
- Order rows by hcluster ordering / draw marker dendrogram| 7

![image](https://user-images.githubusercontent.com/6101677/129816366-fc9746c1-0514-40a5-922e-9b449e3c20b7.png)

y2kbugger commented:
## 2021-09-07

- Fix file not saving where expected (better example in notebook)
- Fix "current metacluster" dropdrop delays reflecting newly added metacluster
- Start debugging "rename current metacluster" glitchiness

y2kbugger commented:
## 2021-09-08

- Finish debugging "rename current metacluster" glitchiness
- Start adding heatmap legend.
- white_check_mark
- eyes
- raised_hands

## 2021-09-09

- Finish adding heatmap legend
- Explore adding hover tooltips

## 2021-09-13

- 50% of input validation done

y2kbugger commented:
Only thing left is to finish validation unless anyone comes back with more feedback, let me know.

ngreenwald commented on ark/utils/metacluster_remap_gui/.gitignore:
```diff
@@ -0,0 +1,3 @@
+Pipfile
```
Do we need a separate .gitignore here?

y2kbugger replied:
I added this rule to ignore files that i personally used to create native venvs (in addition to docker venvs). This running allowed IDE autocompletion, linter rules, pytest, codecoverage, etc.

It's likely this gitignore rule is fine to add to the root `.gitignore`, but I was being conservative.

Additionally I know vscode IDE has support for connecting to docker container envs but it requires at least some configuration. Using a local venv just let me get to work quickly. I could,  work out and document method for running pytest locally via VSCode using the the docker venvs.

ngreenwald replied:
That's okay, I think Adam already has this mostly worked out for his workflow

y2kbugger replied:
Ok. I've removed folder specific .gitignore

ngreenwald commented on .gitignore:
```diff
@@ -21,3 +22,9 @@ env
 */visualizations
+
+data/example_dataset/metaclustering/*output*.csv
+data/example_dataset/metaclustering/example_manually_adjust_metaclusters.ipynb
```
This notebook should get moved to /templates and added to git

y2kbugger replied:
Notebook is already in templates, this was just a quick solution for the opposite of `update_notebooks.sh` from a example notebook developer perspective.

I would love to discuss a better way to to this, honesty my method feels silly, but it worked and was simple.

Use case:
As a developer, I want changes to ipynb files to be immediately reflected in `git status` so that I choose to commit or discard those changes as I commit changes to supporting library files.

My simple solution is shown below, however, it requiree the additional gitignores to prevent tracking the intermediate file.

![image](https://user-images.githubusercontent.com/6101677/134819603-c3c28e8d-fcf8-4163-8f20-b5a3b8411b14.png)

/speculation
Is the reason for copying the templates during docker build something like this:
As data analyst, I don't want any of my changes to ipynb files to be tracked by git so that pulling updates to the notebooks doesn't cause merge conflicts and I can manually pull latest upstream ipynb files using `update_notebooks.sh`

y2kbugger replied:
Just thought of a more robust solution...and it's even simpler.

Add an optional arg to "start_docker.sh" e.g. `--develop-template-notebooks`

Which will bind the templates folder directly. I think this would allow for having the "favor upstream" workflow be the default behavior while also allowing for normal edit-add-commit workflow when developing template notebooks.

ngreenwald replied:
Yes, the idea is that the vast majority of people using the repo are not developers, and we don't want them to run into merge conflicts when they pull new changes. Mounting the templates directly would solve the problem, or working on a copy that's somewhere else until the PR is ready to merge in. @vacuousplanet/ @alex-l-kong  any preference?

ngreenwald commented on ark/utils/metacluster_remap_gui/colormap_helper.py:
```diff
@@ -0,0 +1,29 @@
+    return matplotlib.colors.ListedColormap(rgbs)
+
+
+def distinct_rgbs(n=33):
```
The other helper functions don't need a docstring since they're so short, but I think it would be useful to have one for `distinct_rgbs`

y2kbugger replied:
Agreed  :+1:

ngreenwald commented on ark/utils/metacluster_remap_gui/test_colormap_helper.py:
```diff
@@ -0,0 +1,9 @@
+from .colormap_helper import distinct_cmap, distinct_rgbs
+
+
+def test_colormap_is_distinct():
```
for test functions, please name them original_function_name_test

y2kbugger replied:
After testing this change, pytest does not collect function unless they have a `test_` prefix.

Additionally I checked and it seems the other tests start with `test_` 
https://github.com/angelolab/ark-analysis/search?p=2&q=test
![image](https://user-images.githubusercontent.com/6101677/134821732-f7838baa-50c7-4edb-ae28-cde5cbebf2f9.png)

ngreenwald replied:
Yup, totally right. It should be test_function_name, not function_name_test

ngreenwald commented on ark/utils/metacluster_remap_gui/test_colormap_helper.py:
```diff
@@ -0,0 +1,9 @@
+from .colormap_helper import distinct_cmap, distinct_rgbs
```
for test files, please name them original_file_name_test.py

y2kbugger replied:
Fixed, I like this, makes it sort the test file right next to implementation. Cool to see pytest added this to suffix to file collection in 2018.

ngreenwald commented on ark/utils/metacluster_remap_gui/metaclustergui.py:
```diff
@@ -0,0 +1,463 @@
+import numpy as np
+import ipywidgets as widgets
+import matplotlib.pyplot as plt
+import matplotlib.patheffects as path_effects
```
Please alphabetize these

ngreenwald commented on ark/utils/metacluster_remap_gui/metaclustergui.py:
```diff
@@ -0,0 +1,463 @@
+
+
+class MetaClusterGui():
+    def __init__(self, metaclusterdata, heatmapcolors=sns.diverging_palette(240, 10, n=3, as_cmap=True), width=17.0, debug=False, enable_throttle=True):  # noqa
```
This can be indented instead to avoid the linter errors right?

y2kbugger replied:
Of course, I always try both ways and make a judgement call on readability. One technique to mitigate masking lint error is disabling only one specific lint rather than noqa catchall, pylint supports this for example, but pycodestyle doesn't seem to.

ngreenwald replied:
You didn't like the result of formatting it something like this?
``` 
def __init__ (xxxx, 
            yyy, zzz)
```

y2kbugger replied:
### tldr
I can and will do it this way, I see it is used throughout the rest of the codebase, it's not horrible.

### long

1. Bias against for wrapping any line which starts a block because it (slightly) degrades the block structure and negatively affects "skimmability"
2. Favor wrapping with constant indent of 4 to minimize noise (nearby lines touched) when commit edits. Otherwise renaming the in the first line forces you to adjust multiple lines. It also give you more space when the first line is already long.
    ```
    myobject.my_list_real_long_variable_naming = [
        a,
        b,
        c,
    ]
    ```
    vs
    ```
    myobject.my_list_real_long_variable_naming = [
                                                   a,
                                                   b,
                                                   c,
                                                     ]
    ```

source: https://www.youtube.com/watch?v=x-kB2o8sd5c ~35 minutes in.

ngreenwald replied:
Totally, sounds good

ngreenwald commented on ark/utils/metacluster_remap_gui/metaclustergui.py:
```diff
@@ -0,0 +1,463 @@
+        # initilize data, etc
+        self.update_gui()
+
+    def make_widgets(self):
```
Please add a brief docstring

ngreenwald commented on ark/utils/metacluster_remap_gui/metaclustergui.py:
```diff
@@ -0,0 +1,463 @@
+        self._heatmaps_stale = True
+        self.mcd.save_output_mapping()
+
+    @DEBUG_VIEW.capture(clear_output=False)
```
Do you think the `DEBUG_VIEWs` here are useful to keep in case we need to make changes and debug? Or should they be removed

y2kbugger replied:
They are disabled by default unless turned on. I did build the debug view as a first class feature meant to left in. Toggling is documented in the developer notes here:
![image](https://user-images.githubusercontent.com/6101677/134822116-69189ece-133f-4cd2-9eff-4fece9d59b84.png)

ngreenwald commented on ark/utils/metacluster_remap_gui/metaclusterdata.py:
```diff
@@ -0,0 +1,128 @@
+    assert metacluster_header in clusters.columns, "cluster csv must include column named \"metacluster\", alternately specify the metacluster indexs using keyword `metacluster_index`"  # noqa
+    clusters = clusters.rename(columns={metacluster_header: 'metacluster'})
+    pixelcounts = pd.read_csv(pixelcount_filepath)
+    assert 'cluster' in pixelcounts.columns, "pixelcounts csv must include column named \"cluster\""  # noqa
```
`assert` statements should be reserved for tests only, please `Raise` an error instead for input validation. 

We have a couple helper functions for input validation here: https://github.com/angelolab/ark-analysis/blob/31cc477932b28eba05c3329c807ac2237cf0b143/ark/utils/misc_utils.py#L120

For an example of how they're used see here, or just click on the def in github and go to `references` https://github.com/angelolab/ark-analysis/blob/31cc477932b28eba05c3329c807ac2237cf0b143/ark/mibi/tiling_utils.py#L33

ngreenwald reviewed (changes requested):
Looking good! I left representative comments in a couple places re: test function names, testing patterns, etc, that apply to the rest of the PR as well.

y2kbugger commented on ark/utils/metacluster_remap_gui/test_metaclusterdata.py:
```diff
@@ -0,0 +1,152 @@
+MC_DATA_DIR = DATA_DIR / "example_dataset" / "metaclustering"
+
+
+def test_can_read_example_input_1():
```
This turned out really well. Used my already in place dfs. Mutated the df in place to create the validation issue. Then used it to form an in memory generated csv. This makes it very readable what condition the test is checking for.

    def test_requires_metacluster_column(simple_clusters_df, simple_pixelcount_df):
        simple_clusters_df.rename(columns={'metacluster': 'wrongname'}, inplace=True)
        with pytest.raises(AssertionError):
            metaclusterdata_from_files(as_csv(simple_clusters_df), as_csv(simple_pixelcount_df))

Still have to switch to ValueError so ignore that part.

y2kbugger commented on ark/utils/metacluster_remap_gui/test_data/ex1_clusters_pixelcount_mismatchedids.csv:
```diff
@@ -0,0 +1,100 @@
+cluster,count
```
None now :)

alex-l-kong commented on ark/utils/metacluster_remap_gui/metaclusterdata.py:
```diff
@@ -0,0 +1,128 @@
+        self._metacluster_displaynames_map = {}
+        self._marker_order = list(range(len(self._clusters.columns)))
+
+        assert len(set(self.clusters.index)) == len(list(self.clusters.index)), "Cluster ids must be unique."  # noqa
+        assert set(self.clusters.index) == set(self.cluster_pixelcounts.index), "Cluster ids in both files must match"  # noqa
+        assert 1 in self.clusters.index and 0 not in self.clusters.index, "Cluster ids must be integer, starting with 1."  # noqa
```
Change to `raise ValueError(...)` instead of `assert` statements.

alex-l-kong commented on ark/utils/metacluster_remap_gui/metaclusterdata.py:
```diff
@@ -0,0 +1,128 @@
+        out_df.to_csv(self.output_mapping_filename)
+
+    def set_marker_order(self, new_indexes):
+        assert set(new_indexes) == set(self._marker_order), \
```
Same as above. We actually have a function that does this already: please take a look at `utils.misc_utils.verify_same_elements`.

y2kbugger replied:
So (unlike the others, which i've got on the backlog to fix) this one really is an assert, not input validation. Meant to catch bugs at develop time when someone changes code. Could honestly be removed at this point if your style guide is to have no asserts. let me know.

alex-l-kong commented on ark/utils/metacluster_remap_gui/metaclustergui.py:
```diff
@@ -0,0 +1,463 @@
+            np.zeros(self.mcd.cluster_count))
+        self.labels_cp = []
+        label_alignment_fudge = 0.08
+        for x in np.arange(self.mcd.cluster_count)+0.5+label_alignment_fudge:
```
```suggestion
        for x in np.arange(self.mcd.cluster_count) + 0.5 + label_alignment_fudge:
```

alex-l-kong replied:
I believe our linter will complain that there should be a space in between each value connected by a mathematical operation.

y2kbugger replied:
Not sure why, but pycodestyle doesn't find this case. I'll recheck manually for any others.

alex-l-kong commented on ark/utils/metacluster_remap_gui/metaclustergui.py:
```diff
@@ -0,0 +1,463 @@
+    def preplot(self, df):
+        return df.apply(zscore).clip(upper=self.max_zscore).T
+
+    def make_gui(self):
```
This function is 126 lines: is there a reason we can't break this up into helper functions?

y2kbugger replied:
No i'll break it up. It could also be better factored in terms of "single level of abstraction' principal.

Right now it is a mixture of high and low level. (make_widget, update_gui, vs everything else)
![image](https://user-images.githubusercontent.com/6101677/135541272-4b62029f-f7e6-4f6d-8d93-e63b8631c24c.png)

y2kbugger replied:
I move highlevel coordination to `__init__` and now `make_widgets` and `make_gui`  are now analogous and orthogonal.

alex-l-kong commented on ark/utils/metacluster_remap_gui/metaclustergui.py:
```diff
@@ -0,0 +1,463 @@
+        self.ax_c.yaxis.set_tick_params(which='major', labelleft=False)
+        self.ax_c.set_yticks(np.arange(self.mcd.marker_count)+0.5)
+        self.ax_c.set_yticklabels(self.mcd.marker_names)
+        self.ax_c.set_xticks(np.arange(self.mcd.cluster_count)+0.5)
+        self.ax_m.set_xticks(np.arange(self.mcd.metacluster_count)+0.5)
```
See comment about mathematical operation formatting

alex-l-kong commented on ark/utils/metacluster_remap_gui/metaclustergui.py:
```diff
@@ -0,0 +1,463 @@
+        self.im_m = self.ax_m.imshow(np.zeros((self.mcd.marker_count, self.mcd.metacluster_count)), norm=self.normalizer, cmap=self.heatmapcolors, aspect='auto', picker=True)  # noqa
+
+        self.ax_c.yaxis.set_tick_params(which='major', labelleft=False)
+        self.ax_c.set_yticks(np.arange(self.mcd.marker_count)+0.5)
```
Same comment about mathematical formatting.

alex-l-kong commented on ark/utils/metacluster_remap_gui/metaclustergui.py:
```diff
@@ -0,0 +1,463 @@
+                lb.set_family('monospace')
+                lb.set_zorder(4)
+
+        add_room_for_labels()
+        stretch_dendro_leaves()
+        move_ax_labels()
+        restyle_ax_labels()
```
Any reason we need to do it this way, and we can't do something instead like:

```
# add room for axes labels
ax.set_axisbelow(False)
xlim = ax.get_xlim()
ax.set_xlim((xlim[0], -(xlim[0]*dendrosplit_ratio)))

# stretch dendro leaves
# code in stretch_dentro_leaves

...
```

y2kbugger replied:
Could be done either way, I look into it, might have started out longer than it ended up.

Two clean code heuristics explain why I did it this way.

1. Single level of abstraction
2. Prefer self-documenting function names over comments

## Single level of abstraction
http://principles-wiki.net/principles:single_level_of_abstraction
All statements of a method should belong to the same level of abstraction. If there is a statement which belongs to a lower level of abstraction, it should go to a private method which comprises statements on this level. Doing so will result in smaller methods.

## Prefer self-documenting function names over comments
https://levelup.gitconnected.com/instead-of-a-comment-use-a-well-named-function-448eabb45607

y2kbugger replied:
Ok after sleeping on it I realize you probably asked because I used nested defs. It's to decrease the scope of the def itself (e.g. private methods). Basically its a way of getting benefits of a static helper object with private methods but without overcomplicating architecture.

In languages without nested defs I would have created a separate class with the nested def put in as private methods.

alex-l-kong commented on ark/utils/metacluster_remap_gui/metaclustergui.py:
```diff
@@ -0,0 +1,463 @@
+        self.im_m.set_clim(self.normalizer.vmin, self.normalizer.vmax)
+
+        # xaxis metacluster color labels
+        assert max(self.mcd.metaclusters.index) <= self.mcd.cluster_count, \
+            "Can't support metaclusters idx > cluster count"
```
`raise ValueError(...)` here instead.

alex-l-kong reviewed (commented):
Just a few comments about formatting and code aesthetics.

y2kbugger commented on ark/utils/metacluster_remap_gui/metaclustergui.py:
```diff
@@ -0,0 +1,463 @@
+
+    def make_gui(self):
+        #  |  |    Cluster     | Meta |
+        #  ----------------------------
```
Added better explanation of the Axes map comment.

y2kbugger commented:
Remaining code review items:

- assert -> ValueErrors
- refactor/break out/clarify
    - make_gui
    - move_dendro_labels

y2kbugger commented:
@ngreenwald I have address all previous code review comments from both reviewer. Ready for re-review.

ngreenwald commented on ark/utils/metacluster_remap_gui/metaclusterdata_test.py:
```diff
@@ -0,0 +1,75 @@
+    np.testing.assert_equal(simple_metaclusterdata.metaclusters.values, clusters_data)
+
+
+def test_can_reorder_markers(simple_metaclusterdata: MetaClusterData):
```
These test names as well
